doc/src/manual/ranch.asciidoc

```diff
@@ -207,3 +242,14 @@ connection processes or give them some time to stop properly.
 
 This function does not return until the listener is
 completely stopped.
+
+=== suspend_listener(Ref) -> ok
+
+Ref = ref():: Listener name.
+
+Suspend the given listener if it is running.
+If the listener is already suspended, nothing will happen.
+
+The listener will stop listening and accepting connections by
+closing the listening port, but will not stop running connection
```

Comment:
awesome!

Reply:
ranch:info all_connections key for a number

Reply:
The shutdown strategy you ask for is, to my knowledge, in part not possible.

`SIGTERM` is received and handled by the Erlang VM. Before OTP 19.3, this wasn't even handled and the VM would just die; [since 19.3, `SIGTERM` equals calling `init:stop/0`](http://www.erlang.org/news/110), which will take down all applications and then exit. But the applications don't get any information about why they are being shut down. `SIGTERM` causes `init:stop/0`, which causes `application:stop/1`, and each of those levels is oblivious to why it has been called.

The other part, suspending all listeners on shutdown, incidentially works the way you want if you have only 1 listener. With multiple listeners, there is a catch, see below ;)

The listeners are shut down (in reverse order). For each listener, the `acceptors_sup` (which owns the listening socket) is taken down first, thereby implicitly closing the socket. Then the `conns_sup` is taken down, which will wait for the connection processes to stop. Their allowed shutdown time can be configured with the `shutdown` transport option, iirc it defaults to 5 seconds.

The catch when using multiple listeners stems from the way a supervisor (in this case, `ranch_sup`) shuts down it's children, which means: in sequential fashion (as opposed to in parallel). So only when one listener has stopped, the next is being stopped, which means it was still accepting connections until then.

@essen I explained that right, didn't I? ;) To mitigate this, we could add a `prep_stop/1` in `ranch_app` that calls `ranch:suspend_listener/1` on all listeners? Whether it is wise to do that is up to you to decide.

Reply:
Don't know about wise but it should be in its own ticket.